rpk: group describe supporting regex

daisukebe opened the pull request:
Motivation explained in https://github.com/redpanda-data/redpanda/issues/17710

```
when redpanda admins/support want to get a description of all consumer groups...
the drill is a  "for i in " loop based on "rpk group list" ..
which runs
rpk group describe $i
on each group

We should have the ability to get the group description in a single rpk group command
```

I lean towards supporting regex in the `describe` command as `topic describe` does. As `group describe` already accepts more than one group in one command, I feel it is natural to support regex too.

Fixes https://github.com/redpanda-data/redpanda/issues/17710

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.1.x
- [x] v23.3.x
- [ ] v23.2.x

## Release Notes

### Improvements

* rpk: group describe supports `--regex` flag

r-vasquez commented on src/go/rpk/pkg/cli/group/describe.go:
```diff
@@ -23,14 +25,24 @@ import (
+The --regex flag (-r) parses arguments as regular expressions
+and describes groups that match any of the expressions.
+
+For example,
+
+    describe foo bar            # describe groups foo and bar
+    describe -r '^f.*' '.*r$'   # describe any group starting with f and any groups ending in r
+    describe -r '*'             # describe all groups
+    describe -r .               # describe any one-character groups
```
I know this is similar to what we do in topic delete, but, can we move these examples to the `cobra.Command`'s `Example` field, we are starting to move examples to this field in new commands.

For example: https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cli/cluster/quotas/describe.go#L61-L69

This makes it easier for the doc reference page (example: https://docs.redpanda.com/current/reference/rpk/rpk-security/rpk-security-role-assign/#examples)

daisukebe replied:
Addressed.

r-vasquez commented on src/go/rpk/pkg/cli/group/describe.go:
```diff
@@ -238,3 +259,47 @@ func printLagPerTopic(groups kadm.DescribedGroupLags) {
+	return regexListedGroups(groups.Groups(), expressions)
+}
+
+func regexListedGroups(groups, expressions []string) ([]string, error) {
```
Thanks for splitting these 2 functions, can you write a unit test for this function, please?

daisukebe replied:
Addressed.

r-vasquez reviewed (commented):
LGTM, just 2 requests but thanks for doing this!

daisukebe commented:
Based on @r-vasquez 's feedback, made some changes below and rebased.

* moved examples to the `cobra.Command`'s `Example` field
* made the regex function universal under the `utils` package, `regex_filter.go`
* removed the regex logic from `topic/utils.go`
* moved the test code from `topic/utils_test.go` to `utils/regex_filter_test.go`

micheleRP commented on src/go/rpk/pkg/cli/group/describe.go:
```diff
@@ -17,20 +17,37 @@ import (
+and describes groups that match any of the expressions.
+`,
+		Example: `
+Describe groups foo and bar
+  rpk group describe foo bar
+
+Describe any group starting with f and any groups ending in r
+  rpk group describe -r '^f.*' '.*r$'
+
+Describe all groups
+  rpk group describe -r '*'
+
+Describe any one-character groups
```
```suggestion
Describe groups foo and bar:
  rpk group describe foo bar

Describe any group starting with f and ending in r:
  rpk group describe -r '^f.*' '.*r$'

Describe all groups:
  rpk group describe -r '*'

Describe any one-character group:
```

daisukebe replied:
Thank you!

r-vasquez reviewed (dismissed):
This is awesome, thanks!

daisukebe commented:
/ci-repeat 3

daisukebe commented:
@r-vasquez , is the build failure relevant?

r-vasquez commented:
@daisukebe They were unrelated failures, not sure why the bot didn't retry (I did it manually). Mergin it now, thanks

vbotbuildovich commented:
/backport v24.1.x
